src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -12,179 +12,220 @@
-
     Type *i8ptr_t = PointerType::getUnqual(
         IntegerType::getInt8Ty(mod->getContext()));
-    // The callback to call when a guard fails
-    Constant* gf_handler_ptr = ConstantExpr::getBitCast(
+    // The callback to call when a guard fails.
+    Constant* guardHandler = ConstantExpr::getBitCast(
         mod->getFunction(GUARD_FUN_NAME), i8ptr_t);
     for (auto &bb : fun) {
       for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
-        if (isa<CallInst>(it) &&
-            cast<CallInst>(*it).getCalledFunction()->getName() == "putchar") {
+        if (funName.endswith("more_indirection") && isa<ReturnInst>(it)) {
+          LLVMContext &ctx = bb.getContext();
           IRBuilder<> builder(&bb, it);
-          uint64_t patchpoint_id;
-          if (start_sm_id.find(twin_fun) == start_sm_id.end()) {
-            // Each stackmap ID is calculated by adding the offset
-            // (call_inst_count) to the 'base' ID (curr_sm_id)
-            patchpoint_id = curr_sm_id + call_inst_count;
-          } else {
-            // If the 'twin' of this function has already been processed,
-            // make sure the ID of each stackmap call in this function can be
-            // obtained from an ID of a stackmap call in the 'twin', by
-            // negating the ID. For example, the patch point with ID `0`
-            // corresponds to the patchpoint with ID `~0 = -1`.
-            patchpoint_id = ~(~curr_sm_id + call_inst_count);
-          }
+          uint64_t PPID = getNextPatchpointID(funName);
           // The first two arguments of a stackmap/patchpoint intrinsic call
-          // must be the unique identifier of the call and the number of bytes
-          // in the shadow of the call
-          auto args = std::vector<Value*> { builder.getInt64(patchpoint_id),
-                                            builder.getInt32(13) // XXX shadow
-                                          };
+          // are the unique identifier of the call, and the number of bytes
+          // in the shadow of the call.
+          auto args = vector<Value*> { builder.getInt64(PPID),
+                                       builder.getInt32(13) // XXX shadow
+                                      };
           // The intrinsic to call: patchpoint, if the current function is
-          // trace; stackmap, if the current function is __unopt_trace
+          // optimized; stackmap, if the current function is unoptimized.
           Function *intrinsic = nullptr;
-          if (!fun_name.startswith(UNOPT_PREFIX)) {
-            // The current function is trace -> insert a patchpoint call. The
-            // arguments of the patchpoint call are: a callback, an integer
-            // which represents the number of arguments of the callback, and
-            // the arguments to pass to the callback.
+          if (!funName.startswith(UNOPT_PREFIX)) {
+            // The current function will optimized -> insert a patchpoint call.
```

Comment:
`he current function will optimized`. Seems like a typo?

Reply:
Well, this pass runs 'as early as possible', so at this point in time, both the `__unopt_` version and the regular versions are identical. So really, none of the functions are optimized at this time